File: Week1_Design Patterns and Principles/Ex_1_Implementing_the_Singleton_Pattern/Code/Logger.java


import java.time.LocalDateTime;
import java.time.format.DateTimeFormatter;

public class Logger {

    private static Logger instance;

    private Logger() {
        System.out.println("Logger instance created at: " + getCurrentTimestamp());
    }

    public static synchronized Logger getInstance() {
        if (instance == null) {
            instance = new Logger();
        }
        return instance;
    }

    public void log(String message) {
        System.out.println("[" + getCurrentTimestamp() + "] LOG: " + message);
    }

    public void logError(String errorMessage) {
        System.err.println("[" + getCurrentTimestamp() + "] ERROR: " + errorMessage);
    }

    public void logWarning(String warningMessage) {
        System.out.println("[" + getCurrentTimestamp() + "] WARNING: " + warningMessage);
    }

    public void logInfo(String infoMessage) {
        System.out.println("[" + getCurrentTimestamp() + "] INFO: " + infoMessage);
    }

    private String getCurrentTimestamp() {
        return LocalDateTime.now().format(DateTimeFormatter.ofPattern("yyyy-MM-dd HH:mm:ss"));
    }

    public String getInstanceDetails() {
        return "Logger instance: " + this.hashCode();
    }
}

File: Week1_Design Patterns and Principles/Ex_1_Implementing_the_Singleton_Pattern/Code/Logger.cs
using System;

namespace SingletonDemo;


public sealed class Logger
{
    private static Logger? _instance;
    private static readonly object _lock = new();

    private Logger()
    {
        Console.WriteLine($"Logger instance created at: {GetCurrentTimestamp()}");
    }

    public static Logger Instance
    {
        get
        {
            if (_instance is null)
            {
                lock (_lock)
                {
                    _instance ??= new Logger();
                }
            }
            return _instance;
        }
    }

    public void Log(string message)       => Console.WriteLine($"[{GetCurrentTimestamp()}] LOG: {message}");
    public void LogError(string message)  => Console.Error.WriteLine($"[{GetCurrentTimestamp()}] ERROR: {message}");
    public void LogWarning(string message)=> Console.WriteLine($"[{GetCurrentTimestamp()}] WARNING: {message}");
    public void LogInfo(string message)   => Console.WriteLine($"[{GetCurrentTimestamp()}] INFO: {message}");

    public string InstanceDetails => $"Logger instance: {GetHashCode()}";
    private static string GetCurrentTimestamp() => DateTime.Now.ToString("yyyy-MM-dd HH:mm:ss");
}
